### ISSUE TITLE
unable to install pyside with nix on osx sierra

### ISSUE THREAD

**Issue by javier-martinez-palmer** [NixOS/nixpkgs#25619]
## Issue description

pyside is not compiling for pythin2.7

### Steps to reproduce

I am new to nix, but got it working for imagemagick and other packages. 
I got python, pip and wheel installed but if I try to install pyside using this command:
`nix-env -i python2.7-pyside`
I got a compiling error. I would expect all dependencies being downloaded and compiled.

The error looks like this:

```
1 warning generated.
4 warnings generated.
[  3%] Linking CXX shared library libpyside-python2.7.dylib
[  3%] Built target pyside
make: *** [Makefile:139: all] Error 2
builder for ‘/nix/store/8q7pafp19jzhaam07dj17rh5m5na12vc-python2.7-pyside-1.2.4.drv’ failed with exit code 2
error: build of ‘/nix/store/8q7pafp19jzhaam07dj17rh5m5na12vc-python2.7-pyside-1.2.4.drv’ failed

```
Nix looks awesome so far. Some help would be great, thanx in advance.

## Technical details

* System: osx sierra 10.12.4
* Nix version: nix-env (Nix) 1.11.9
* Nixpkgs version: 17.09pre107131.c882b967c1
* Sandboxing enabled:  /etc/nix/nix.conf: No such file or directory

**Comment by FRidh** [comment]
That error of course shouldn't happen. I'm not using Darwin so I can't help you much further here.

In any case, `nix-env -i python2.7-pyside` is not a supported workflow for Python on Nix. Please read the Python section in the Nixpkgs to find out how to install a Python environment.

**Comment by matheboy** [comment]
Well, not sure if I did it right, most documentation seems to be outdated, but among other blogs and documentation I found this https://nixos.org/wiki/Python
Those are the packages I have installed when I run nix-env -q:
imagemagick-6.9.7-6
nix-1.11.9
nss-cacert-3.30
p7zip-16.02
python-2.7.13
qt-4.8.7
( qt-4.8.7 should not be needed, probably is detected as dependency already but I installed it just in case  ) 
When I do nix-env -qaP | grep pyside I get this :
nixpkgs.pysideApiextractor                                     pyside-apiextractor-0.10.10
nixpkgs.pysideGeneratorrunner                                  pyside-generatorrunner-0.6.16
nixpkgs.python27Packages.pyside                                python2.7-pyside-1.2.4
nixpkgs.python27Packages.pysideShiboken                        python2.7-pyside-shiboken-1.2.4
nixpkgs.python27Packages.pysideTools                           python2.7-pyside-tools-0.2.15

but as said, if I do nix-env -i python2.7-pyside or nix-env -i python2.7-pyside-1.2.4 doesnt work...

**Comment by copumpkin** [comment]
Here's the actual error: 

```
/private/var/folders/vv/17b_hycx691699h20v458fnh0000gn/T/nix-build-python2.7-pyside-1.2.4.drv-0/PySide-1.2.4/plugins/customwidget.h:27:10: fatal error: 'QtDesigner/QtDesigner' file not found
#include <QtDesigner/QtDesigner>
         ^~~~~~~~~~~~~~~~~~~~~~~
```

To me that suggests that on Darwin, Qt is supposed to produce a .framework, and perhaps our common Qt builder doesn't build that. Someone more Qt-focused might want to chime in. Perhaps @periklis?

**Comment by matheboy** [comment]
Installing pip, wheel, setuptools and qt and using pip to install pyside gives the same error. In case this helps anyone. 

As a side note, I also had to fix some rights under the /nix folder. Suddenly I got no rights to write or remove some files ( nix-collect-garbage -d. and nix-env -i were failing because of that)

**Comment by periklis** [comment]
If you need the QtDesigner Headers then you should add `qttools` to your buildInputs. I patched a while ago the submodule qttools for qt-5/5.8 to work on darwin.

**Comment by matheboy** [comment]
I don't need them. This is for pyside that is using qt-4.8.7.

**Comment by periklis** [comment]
But pyside needs qt and specifically qtdesigner from qttools. I am not sure if we even build qttools for 4.8.7

**Comment by matheboy** [comment]
Yeah well I don't mind having the designer or other tools, but there is nothing I specifically need. With or without attools I don't mind. 
That being said, if you need them for compiling pyside or qt you should solve that dependency...
In fact those may be gathered already (probably not the right version?)interesting test this, but sorry I cannot test right now

**Comment by matheboy** [comment]
should be simple to fix this just the same for 4.8.7 right?

**Comment by periklis** [comment]
Oh boy, i don't understand a word of the last two comments. However, let me explain it again. In order to compile the package pyside you need qtdesigner, which i suspect that we would do not provide for qt 4.8.7 for darwin right now. What's next?
Please try to build qt-4.8.7 with nix-build on darwin and check if we you get qtdesigner. 

Imho the only qt version in nixpkgs providing the qttools/qtdesigner stuff is qt5/5.8.

**Comment by matheboy** [comment]
I think we understand each other very well. 
I assume as well pttools are not working in qt-4.8.7. 
The problem is I don't have the time to test this for you.

**Comment by Enzime** [comment]
Pretty sure the tools are being built, as Qt Designer can be found at `${qt4}/bin/Designer.app/Contents/MacOS/Designer` and successfully runs.

After changing `makeFlags` in `pkgs/development/python-modules/pyside/default.nix` to 
```makeFlags = [ "QT_PLUGIN_PATH=${pysideShiboken}/lib/generatorrunner" "VERBOSE=1" ];```

I get the following log which includes the commands that `make` is running:

https://gist.github.com/Enzime/d24cc4910e0ddf02cba1fc43078b0e27

Of note is line 625, which has the command for compiling `plugins/customwidgets.cpp` the file where it fails to find `<QtDesigner/QtDesigner>`.

On this line you can find a clearly invalid path `-iframework /nix/store/nd412ggfppsh7zwdml221psjkzznvycf-qt-4.8.7/include/nix/store/nd412ggfppsh7zwdml221psjkzznvycf-qt-4.8.7/lib`.

After adding `cmakeFlags = [ "-DALTERNATIVE_QT_INCLUDE_DIR=${qt4}/lib/QtCore.framework" ];` to to the `pyside/default.nix`, this leads to it correctly(?) generating `-iframework /nix/store/nd412ggfppsh7zwdml221psjkzznvycf-qt-4.8.7/lib` meaning that it now successfully finds `<QtDesigner/QtDesigner>` however the compilation fails later...

https://gist.github.com/Enzime/5a1e844358970533b8e11e1fbca1fc37

**Comment by Enzime** [comment]
After noticing `shiboken` looking in the wrong directory for headers e.g. `/nix/store/nd412ggfppsh7zwdml221psjkzznvycf-qt-4.8.7/lib/QtCore.framework/QtCore.framework/Headers/qglobal.h` (line 516 in the previous log) I realized this was related to setting the `ALTERNATIVE_QT_INCLUDE_DIR` to `${qt4}/lib/QtCore.framework` instead of just `${qt4}/lib`.

After changing `cmakeFlags` I realized this now leads to the library being included as a system include dir instead of a framework include dir (`-isystem` as opposed to `iframework`) due to the Qt4 CMake Macros, so I had to add `"-DCMAKE_CXX_FLAGS=-F${qt4}/lib"` to `cmakeFlags` to get it to read the dir as a framework dir, effectively rendering the `-isystem /nix/store/nd412ggfppsh7zwdml221psjkzznvycf-qt-4.8.7/lib` generated by cmake useless.

This allows me to successfully compile `pyside` as well as `syncplay`.

However, whenever I attempt to run `syncplay` this leads to a segfault, potentially related to #31865 :/

https://gist.github.com/Enzime/ab6b6fd25581e87f062bfcf31146cb45

If anyone has another way of verifying that my instructions lead to a valid `pyside` compilation, I'd be happy to test.
